Bottom: 44a177b983999132877889e122ebf4a84d135ffa
Top:    8d5203b9734e69cd1f05ebedd5c8d3ab5160d2a6
Author: Ben Greear <greearb@candelatech.com>
Date:   2015-03-31 14:07:12 -0700

ath10k: Allow max-msdu to be configured with mod-param.

The 4.0 kernel is using more firmware RAM, probably because
of larger htc buffers.  To still support 64 vdevs, I need
to save a bit of RAM.  Do that by decreasing the default
value for msdu-desc count.  It may be adjusted with module
parameter as desired by the user.

Signed-off-by: Ben Greear <greearb@candelatech.com>


---

diff --git a/drivers/net/wireless/ath/ath10k/core.c b/drivers/net/wireless/ath/ath10k/core.c
index 8fd42a0..6351afd 100644
--- a/drivers/net/wireless/ath/ath10k/core.c
+++ b/drivers/net/wireless/ath/ath10k/core.c
@@ -1595,7 +1595,7 @@ static int ath10k_core_init_firmware_features(struct ath10k *ar)
 			ar->max_num_peers = ath10k_modparam_target_num_peers_ct;
 			ar->max_num_stations = TARGET_10X_NUM_STATIONS;
 			ar->max_num_vdevs = ath10k_modparam_target_num_vdevs_ct;
-			ar->htt.max_num_pending_tx = TARGET_10X_NUM_MSDU_DESC_CT;
+			ar->htt.max_num_pending_tx = ath10k_modparam_target_num_msdu_desc_ct;
 		} else {
 			ar->max_num_peers = TARGET_10X_NUM_PEERS;
 			ar->max_num_stations = TARGET_10X_NUM_STATIONS;
diff --git a/drivers/net/wireless/ath/ath10k/hw.h b/drivers/net/wireless/ath/ath10k/hw.h
index e2f01fa..80621f9 100644
--- a/drivers/net/wireless/ath/ath10k/hw.h
+++ b/drivers/net/wireless/ath/ath10k/hw.h
@@ -376,8 +376,6 @@ enum ath10k_hw_4addr_pad {
 						    * override w/module parm */
 #define TARGET_10X_AST_SKID_LIMIT_CT		(ath10k_modparam_target_num_peers_ct * TARGET_10X_NUM_PEER_AST)
 #define TARGET_10X_NUM_PEER_KEYS_CT             (WMI_MAX_KEY_INDEX + 1) /* 4 */
-/* These eat a fair chunk of memory on the firmware, so decrease it a bit. */
-#define TARGET_10X_NUM_MSDU_DESC_CT		808 /* must be multiple of 8 */
 
 /* Related to HTC buffers */
 /* return any credit immediately */
diff --git a/drivers/net/wireless/ath/ath10k/mac.c b/drivers/net/wireless/ath/ath10k/mac.c
index 70c60ba..4cd7de1 100644
--- a/drivers/net/wireless/ath/ath10k/mac.c
+++ b/drivers/net/wireless/ath/ath10k/mac.c
@@ -191,6 +191,11 @@ int ath10k_modparam_target_num_peers_ct = 128;
 module_param_named(num_peers_ct, ath10k_modparam_target_num_peers_ct, int, 0444);
 MODULE_PARM_DESC(num_peers_ct, "Maximum peers to request from firmware");
 
+/* These consume a fair bit of RAM on target. */
+int ath10k_modparam_target_num_msdu_desc_ct = 680;
+module_param_named(num_msdu_desc_ct, ath10k_modparam_target_num_msdu_desc_ct, int, 0444);
+MODULE_PARM_DESC(num_msdu_desc_ct, "Maximum MSDU Descriptors in firmware (must be multiple of 8)");
+
 /**********/
 /* Crypto */
 /**********/
diff --git a/drivers/net/wireless/ath/ath10k/mac.h b/drivers/net/wireless/ath/ath10k/mac.h
index d79369a..ddd5b1c 100644
--- a/drivers/net/wireless/ath/ath10k/mac.h
+++ b/drivers/net/wireless/ath/ath10k/mac.h
@@ -29,6 +29,7 @@ enum wmi_tlv_tx_pause_action;
 extern int ath10k_modparam_nohwcrypt;
 extern int ath10k_modparam_target_num_vdevs_ct;
 extern int ath10k_modparam_target_num_peers_ct;
+extern int ath10k_modparam_target_num_msdu_desc_ct;
 
 struct ath10k_generic_iter {
 	struct ath10k *ar;
